[MRG] Transform GP-optimizer space to [0, 1]

MechCoder opened the pull request:
Also,

1. Created a Pipeline transformer
2. Added `HyperCube` transformer to the list of transformers.

MechCoder commented:
@glouppe @betatim please review.

MechCoder commented:
Not sure, why the convergence tests pass in Python3 but fail in Python2.

betatim commented:
(restarted travis in the hopes that it is random ... lame solution I know)

betatim commented:
Not sure I like the new `transform_space` parameter. Especially as parameter for a `Dimension` I think it should just be `transform`.

I like having a chain of transformers.

The other thing that is ugly is passing `dimensions` and a `space` to `base_minimize`. If we really need the space, then we should exclusively pass that as it contains the dimensions.

Some questions I am asking myself (my answers in brackets):
* GP always wants to transform to `(0, 1)`? (yes)
* should the user be able to override that choice? (no)
* other minimisation methods don't want things to be transformed to `(0,1)` or want to apply their own transforms to some of the dimensions (correct)
* does the user ever need to be able to specify their own transform? (maybe, support this by allowing people to pass a fully configured `Space` instance to `dimensions=`)

-> Have a `transform` parameter for `Dimension` that GP sets to `Normalized()` for those that it wants to have normalised.

WDYT?

betatim commented:
Might be worth making [this code](https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/space.py#L350-L369) available as `check_dimension` or something like that so that `gp_minimize` (and friends) can use it to take user specified arguments and turn them into a `Dimension`.

MechCoder commented:
> Especially as parameter for a Dimension I think it should just be transform.

That was my idea as well, but then there is a `transform` method as well. So I cannot set `self.transform = transform`. We can underscore that particular attribute then.

I have to have a closer read, but I agree to all the other points.

MechCoder commented:
I've introduced `check_dimensions` that has an optional `transform` parameter because I preferred that to the idea of setters and getters on Dimension instances. For now I've assumed that if `transform` is provided all dimensions have to be transformed. If needed, later we can allow `transform` to be a list of strings as well.

I've also moved the transformers to a separate file.

betatim commented:
> That was my idea as well, but then there is a `transform` method.

Should we name the argument `transformation=`? Can already see myself forever trying to remember which of `sp.transform_` and `sp.transform` is the attribute and which the method :)

betatim commented on skopt/optimizer/gp.py:
```diff
@@ -3,15 +3,15 @@
+from skopt.optimizer.base import base_minimize
+from skopt.learning import GaussianProcessRegressor
+from skopt.space import check_dimensions
+from skopt.space import Space
```
relative imports are out of fashion? I think we should stick with using them, or make a new PR where we switch the whole code base.

glouppe replied:
Nope. We should keep relative imports.

betatim commented on skopt/space.py:
```diff
@@ -5,98 +5,70 @@
-        return [
-            self.inverse_mapping_[i] for i in self._lb.inverse_transform(Xt)
-        ]
+def check_dimensions(dimensions, transform=None):
```
Originally thought about having `check_dimension()` (no s) and the loop being performed by the caller. It would make it easier for the caller to decide what transform to they wanted for what dimension. Not sure if that is needed or over engineering. You decide.

betatim commented on skopt/space.py:
```diff
@@ -5,98 +5,70 @@
+        - an instance of a `Dimension` object (`Real`, `Integer` or
+          `Categorical`).
+
+    * `transform` ["hypercube", optional] :
```
Nitpicking ahead: I really dislike `'hypercube'` as name for this 😃 `'unithypercube'` maybe, but that is super long. Why not boring old `'normalize'`?

glouppe replied:
+1 for nomalize

MechCoder replied:
I was hesitant because `normalize` can mean a number of things. Is there a standard way to define this scaling?

betatim commented on skopt/space.py:
```diff
@@ -205,9 +198,10 @@ def bounds(self):
-        if self.prior == "uniform":
+        if self.transform_ == "hypercube":
+            return (0.0, 1.0)
+        elif self.prior == "uniform":
```
Had to look at this a few times and still not sure if mixing `self.prior` and `self.transform_` in the `if/elif` does what we want. I'd prefer splitting these into two different `if` blocks one for priors, one for transforms.

betatim commented on skopt/tests/test_space.py:
```diff
@@ -225,3 +227,38 @@ def test_space_api():
+    X_scaled = 28 * (X - X.min()) / (X.max() - X.min()) + 2
+    X_scaled = np.clip(X_scaled, 2.01, 30.01)
+
+    # Check transform
```
-> Check transformed values are all between zero and one

(because both the asserts are _less(...) it took me a while to parse it, hopefully the comment will help with that)

betatim replied:
Do we want a transform that does `0<= x <= 1` or `0 < x < 1`? The test here is for the latter but my guess would be that we want out transform to be the first?

Maybe instead of random values or in addition to them we should test explicitly the edge cases.

betatim commented on skopt/transformers.py:
```diff
@@ -0,0 +1,134 @@
+        self.low = low
+        self.high = high
+
+    def transform(self, X):
```
Need some error handling for when the user wants to transform a value outside of the `low -> high` range.

betatim commented on skopt/space.py:
```diff
@@ -5,98 +5,70 @@
+from skopt.transformers import HyperCube
+from skopt.transformers import Identity
+from skopt.transformers import Log10
+from skopt.transformers import Pipeline
```
Same comment about relative vs absolute imports.

glouppe commented:
Thanks Manoj! I need to have a closer look, I'll try to find some spare time for a review later today.

glouppe commented:
From a code organization point of view, what do you think of defining a transformation as a wrapper around a `Dimension` object (i.e. as function taking a dimension as input and producing a new dimension as output)? In this way, chaining transformations would be done in a functional way (e.g. `Hypercube(Real(0, 10))`). 

(not strong opinion yet, I am thinking aloud...)

glouppe commented on skopt/optimizer/base.py:
```diff
@@ -168,9 +168,9 @@ def base_minimize(func, dimensions, base_estimator,
             "%s has to be a regressor." % base_estimator)
 
+    space = Space(dimensions)
     # Check params
```
this comment should be moved above

glouppe commented on skopt/space.py:
```diff
@@ -5,98 +5,70 @@
         return '...'
 
 
-class _Identity:
```
Not sure I like moving everything to a top-level `transformers` module. Maybe create instead a `space/` submodule?

MechCoder commented:
> From a code organization point of view,

If we do that, then it might make sense to completely decouple the transformations from the dimension instances. Something like just defining the `init` logic in Dimensions and the `transform`, `rvs` attributes to the Transformer.

```python
class Real(Dimension):
    def __init__(self, low, high):
        self.low = low
        self.high = high

class Log10(Transformer):
    def __init__(self, dimension):
        self.dimension = dimension
        self._rvs = uniform(
            np.log10(self.low),
            np.log10(self.high) - np.log10(self.low))

    def rvs(self, n_samples):
        # Sample from rvs

    def transform(self, X):
        np.log10
```        

and then call `Log10(Real(10**-5, 10**5))` internally.

MechCoder commented:
What do we do about a chain of transformations?

```
Pipeline([transformers], Dimension)
```

would be a straightforward way.

MechCoder commented:
I like the idea, because it seems cleaner to me.

glouppe commented:
You could similarly do TransformB(TransformA(BaseDimension)) since
TransformA would share the Dimension API.

On 29 September 2016 at 16:07, Manoj Kumar <notifications@github.com> wrote:

> I like the idea, because it seems cleaner to me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-optimize/scikit-optimize/pull/237#issuecomment-250476066>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdKS-j5Di5mxQVFyi2NTDEehFj2C5BOks5qu8YPgaJpZM4KH7Xw>
> .
>

MechCoder commented:
I would need some time to think but is there a particular reason that you would prefer chained function transformers as opposed to a class accepting a list of transformers and a Dimension instance?

glouppe commented:
The way I see a `Dimension` object is basically as a set and a transformation as a function taking a set as input and producing another set as output. Or more formally, if we have a transformation `f` and a dimension `D`, then `f(D)` is the image of D through `f`, and is itself a set/dimension. This is why I think it would be natural to define transformation as a wrapper object taking as argument another Dimension object and then implementing the `Dimension` interface. Another argument, is that, in my opinion, it might make things cleaner in terms of composition if a `Dimension` does not have to care about how it is transformed.

glouppe commented:
(Of course this is all open for discussion, but we should really think twice on this important issue)

MechCoder commented:
The more I think about it, the more it seems to make sense that way.

MechCoder commented:
@glouppe @betatim In your example, that is `Normalize(Real(0, 10))` or more generally `TransformB(TransformA(BaseDimension))` after sampling from the resultant dimensional space, there is no way to go back to the original dimensional space for a given sample, unless you have implemented an inverse mapping as well. 

Which is why I feel, the current interface is all right, i.e for a Dimension:

1. `rvs` samples it from the original space
2. `transform` maps it to a transformed space as given by the `transformer` argument.
3, `inverse_transform` maps it to from the transformed space to the original space.

Alternatively, we could do away with the `transform` and `inverse_transform` on the dimension objects and replace it with a `Transformer` object that takes in the original space and warped space as class arguments, but I would not prefer that.

glouppe commented:
Is this now superseded by #243 ?

glouppe commented:
> or more generally TransformB(TransformA(BaseDimension)) after sampling from the resultant dimensional space, there is no way to go back to the original dimensional space for a given sample, unless you have implemented an inverse mapping as well.

Why wouldnt we implement the inverse mapping as well?

MechCoder commented:
Let's say I implement a Transformer as a function that returns a new dimensional space and I apply a  chain of transformations to convert the original Dimensional space (provided by the user) to the new Dimensional space (supplied to the optimizer). I also implement the inverse mapping, i.e chain of transformations to go back to the original dimensional space. Also now I have a new point as computed by the optimizer.

How do you suggest to convert this new point to the original dimensional space? (The inverse mapping takes in a Dimension and returns a Dimension)

glouppe commented on skopt/space/space.py:
```diff
@@ -5,98 +5,65 @@
-        return [
-            self.inverse_mapping_[i] for i in self._lb.inverse_transform(Xt)
-        ]
+def check_dimension(dimension, transform=None):
```
Can you also document about the checks that are done?

glouppe commented on skopt/space/space.py:
```diff
@@ -163,35 +130,53 @@ def __init__(self, low, high, prior="uniform"):
+                    [Identity(), Normalize(low, high)])
+            else:
+                self.transformer = Pipeline(
+                    [Log10(), Normalize(np.log10(low), np.log10(high))]
```
This is OK for now, but this is why the functional approach is better in my opinion. For example in this case, the second step in the pipeline need to know the bounds from the previous step, which is not that clean.

MechCoder replied:
But how to address this? (https://github.com/scikit-optimize/scikit-optimize/pull/237#issuecomment-251415962)

glouppe commented on skopt/space/space.py:
```diff
@@ -163,35 +130,53 @@ def __init__(self, low, high, prior="uniform"):
         self.low = low
         self.high = high
         self.prior = prior
+        self.transform_ = transform
```
is there a reason to use a trailing underscore here?

glouppe replied:
Oh right, we already have the `transform` method. Yuck.

glouppe commented on skopt/tests/test_space.py:
```diff
@@ -225,3 +227,35 @@ def test_space_api():
         assert_array_equal(b1, b2)
+
+
+def test_real_hypercube():
```
This test should be renamed

glouppe commented on skopt/tests/test_gp_opt.py:
```diff
@@ -21,7 +21,7 @@ def check_minimize(func, y_opt, bounds, acq_optimizer, acq_func,
 
 def test_gp_minimize():
-    for search, acq in product(["sampling", "lbfgs"], ["LCB", "PI", "EI"]):
+    for search, acq in product(["sampling", "lbfgs"], ["LCB", "EI"]):
```
what is the reason for removing PI?

MechCoder replied:
I removed it because tests were failing in Python2 by a small value and I did not want to increase the number of iterations for that. Moreover, I'm not sure PI is commonly used.

glouppe commented:
Not super fan of the API, but that will do the job for now. +1 for merge once the other comments are addressed.

glouppe commented on skopt/space/__init__.py:
```diff
@@ -0,0 +1 @@
+from .space import *
```
Can you add a top-level docstring for the module?

MechCoder commented:
@glouppe Addressed your comments, Anything else?
